include/glow/Graph/TensorLayout.h

```diff
@@ -0,0 +1,194 @@
+  /// Tensor dimensions descriptions for all dimensions.
+  std::string dims_[max_tensor_dimensions];
+  /// The serialization of the layout.
+  std::string serializedLayout_;
+  /// Expected number of dimensions.
+  size_t numDims_;
+
+public:
+  virtual ~TensorLayoutDescription() = default;
+  /// Constructs this helper class from a serialized string representation.
+  TensorLayoutDescription(std::string text);
+  /// Constructs this helper class from an array of strings representing each
+  /// individual / pre-separated dimension.
+  TensorLayoutDescription(llvm::ArrayRef<std::string> dims);
+  /// \returns the alignment of a dimension \p n.
+  size_t getAlignment(size_t n) const;
+  /// \returns the alignment by parsing dimension string \p s.
+  size_t getAlignment(const std::string &s) const;
+  /// \returns true if both tensor layouts are the same.
+  bool isSameLayout(const TensorLayoutDescription &rhs) const;
+  /// \returns description of the dimension \p n.
+  const llvm::StringRef getNthDimDescription(size_t n) const;
+  /// \returns the description of all dimensions.
+  llvm::ArrayRef<std::string> getDims() const;
+  /// \returns number of dimensions.
+  size_t getNumDims() const { return numDims_; }
+  /// \returns layout name.
+  llvm::StringRef getSerializedLayout() const { return serializedLayout_; }
+  /// \returns true if the layout is "*" in all dimensions.
+  bool isAnyLayout();
+  std::string getDebugDesc() const;
+
+protected:
+  /// parse helper: get the custom extensions information. the default, virtual,
+  /// implementation just ignores all the data until the end token.
+  virtual void parseCustomExtensions(llvm::StringRef &text, unsigned idx);
+
+private:
+  /// Constructor helper: Parses the  serialized string.
+  void parse(llvm::StringRef text);
+
+  /// parse helper: get the official extensions information.
+  void parseOfficialExtensions(llvm::StringRef &text, unsigned idx);
+};
+
+/// Interface for finding out layout requirements.
+class TensorLayoutCommon {
+public:
+  /// \return the default n-D layout for Glow.
+  virtual std::string getDefaultNDLayout(unsigned dims) const;
+
+  /// \returns layout requirements of the Nth input \p n of a Node \p node.
+  virtual std::string getNthInputLayoutRequirements(const Node *node, size_t n);
+
+  /// \returns layout requirements of the Nth result \p n of a Node \p node.
+  virtual std::string getNthResultLayoutRequirements(const Node *node,
+                                                     size_t n);
+
+  /// \returns true if type \p ty satisfies the \p destLayout layout. If \p
+  /// srcLayout is provided, it is taken into account as well.
+  virtual bool isSatisfiedBy(TypeRef ty,
+                             const TensorLayoutDescription &destLayout,
+                             const TensorLayoutDescription *srcLayout) const;
+
+  /// \return layouts for all tensor dimensions.
+  virtual std::array<TensorLayoutDescription, max_tensor_dimensions + 1> &
```

Comment:
Why `std::array` and not e.g. `llvm::ArrayRef`?

Reply:
The g++ 5 compiler does not like that, even though clang/llvm has no problem with it. complains that it cannot convert the c-style array (if we remove std::array) into llvm's ArrayRef

Reply:
@shajrawi Could you use some kind of `llvm::makeArrayRef` here to convert `layoutsForDims` into an `ArrayRef`?

Reply:
Done